returnn/flow.py

```diff
@@ -0,0 +1,122 @@
+        <link from="tf-fwd:log-posteriors" to="network:features"/>
+
+    With the config settings:
+
+        [flf-lattice-tool.network.recognizer.feature-extraction.tf-fwd.input-map.info-0]
+        param-name             = input
+        seq-length-tensor-name = extern_data/placeholders/<feature_tensor_name>/<feature_tensor_name>_dim0_size
+        tensor-name            = extern_data/placeholders/<feature_tensor_name>/<feature_tensor_name>
+
+        [flf-lattice-tool.network.recognizer.feature-extraction.tf-fwd.loader]
+        meta-graph-file    = <tf_graph>
+        required-libraries = <native_ops>
+        saved-model-file   = <tf_checkpoint>
+        type               = meta
+
+        [flf-lattice-tool.network.recognizer.feature-extraction.tf-fwd.output-map.info-0]
+        param-name  = <output_type>
+        tensor-name = <output_tensor_name>/output_batch_major
+
+
+    :param tf_graph: usually the output of a CompileTFGraphJob
+    :param tf_checkpoint: the checkpoint to load the model from, e.g. a ReturnnTrainingJob or similar
+    :param output_type: type of the returned values, see RASR documentation for possible types
+    :param feature_tensor_name: name of the extern data ebtry to feed the features to
+    :param output_tensor_name: the name of the output tensor, it is expected that "<name>/output_batch_major" exists
+    :param native_ops: list of native op ".so" files to link
+    :param tf_fwd_input_name: Naming for the tf network input, usually no need to be changed
+    :return: tensorflow-forward node flow with output link and related config
+    """
+
+    # tf flow (model scoring done in tf flow node) #
+    tf_flow = rasr.FlowNetwork()
+    tf_flow.add_input(tf_fwd_input_name)
+    tf_flow.add_output("features")
+    tf_flow.add_param("id")
+
+    tf_fwd = tf_flow.add_node("tensorflow-forward", "tf-fwd", {"id": "$(id)"})
+    tf_flow.link(f"network:{tf_fwd_input_name}", tf_fwd + ":input")
+    tf_flow.link(tf_fwd + ":" + output_type, "network:features")
+
+    tf_flow.config = rasr.RasrConfig()
+    tf_flow.config[tf_fwd].input_map.info_0.param_name = "input"
+    tf_flow.config[
+        tf_fwd
+    ].input_map.info_0.tensor_name = (
+        f"extern_data/placeholders/{feature_tensor_name}/{feature_tensor_name}"
+    )
+    tf_flow.config[tf_fwd].input_map.info_0.seq_length_tensor_name = (
+        f"extern_data/placeholders/"
+        f"{feature_tensor_name}/{feature_tensor_name}_dim0_size"
+    )
+
+    tf_flow.config[tf_fwd].output_map.info_0.param_name = output_type
+    tf_flow.config[
+        tf_fwd
+    ].output_map.info_0.tensor_name = f"{output_tensor_name}/output_batch_major"
+
+    tf_flow.config[tf_fwd].loader.type = "meta"
+    tf_flow.config[tf_fwd].loader.meta_graph_file = tf_graph
+    tf_flow.config[tf_fwd].loader.saved_model_file = tf_checkpoint
+    if native_ops is not None:
+        tf_flow.config[tf_fwd].loader.required_libraries = native_ops
```

Comment:
Does this work with a list? Right now I only see one case in the recipes where a list of ops is used and there, `":".join()` is applied to it before passing it.

Reply:
no, this will not work as the default stringification is space-separated-list but required_libraries requires colon-separated-list